gain_scheduling.jl

```diff
@@ -0,0 +1,81 @@
+using ControlSystems, OrdinaryDiffEq
+
+"""
+    GainSchedulingSimulator
+# Fields:
+    s::Simulator
+    f = (x,p,t) -> x
+    y = (x,t) -> y
+    r = (x,t) -> reference
+    e = (x,t) -> r(x,t) .- x # Error
+    controllers::Vector{StateSpace}
+    conditions::Vector{Function}
+"""
+struct GainSchedulingSimulator <: ControlSystems.AbstractSimulator
+    s::Simulator
+    f
+    y
+    r
+    e
+    controllers::Vector{StateSpace}
+    conditions::Vector{T} where T <: Function
+    feedforward::Vector{Float64}
+end
+"""
+    GainSchedulingSimulator(P,r,controllers::AbstractVector{LTISystem},conditions::AbstractVector{Function (x,y,r)->Bool}; inputfun=(u,t)->u, feedforward=zeros())
+`r` is a function (x,t)-> reference signal
+`controllers` is a list of LTIsystems representing the controllers
+`conditions` is a list of function that accepts state, output and reference and returns a Bool indicating which controller is active. At least one function should return true for every input. If more than one condition is true, the index of the first condition in the list is chosen.
+´feedforward´ is a vector of constants that are added to the control signal calculated by the active controller, i.e., `u = C*x + D*e + feedforward[active_index]`
+For a usage example, see notebook in example folder.
+```
+"""
+function GainSchedulingSimulator(P,ri,controllers::AbstractVector{Tu},conditions::AbstractVector{Tc};
+    inputfun=(u,t)->u,
+    feedforward=zeros(length(controllers))) where Tu <: StateSpace where Tc <: Function
+    s = Simulator(P)
+    pinds = 1:P.nx # Indices of plant-state derivative
+    r = (x,t) -> ri(x[pinds],t)
+    y(x,t) = s.y(x[pinds],t)
+    y(sol::ODESolution,t) = P.C*sol(t)[pinds,:]
+    e = (x,t) -> r(x,t) .- y(x,t)
+    f = function(der,x,p,t)
+        xyr = (x[pinds],y(x,t),r(x,t))
+        index = findfirst(c->c(xyr...), conditions)
+        @assert index > 0 "No condition returned true"
```

Comment:
findfirst returns `nothing` instead of 0 in Julia 1

Reply:
You're right, good catch - I haven't tested the code beyond the examples in the notebook, so there may be more such errors.